LLama/LLamaContext.cs

```diff
@@ -45,17 +60,68 @@ public class LLamaModel: IDisposable
         /// <summary>
         /// 
         /// </summary>
-        /// <param name="Params">Model params.</param>
+        /// <param name="params">Model params.</param>
         /// <param name="encoding">Encoding to deal with text input.</param>
         /// <param name="logger">The logger.</param>
-        public LLamaModel(IModelParams Params, string encoding = "UTF-8", ILLamaLogger? logger = null)
+        [Obsolete("Use the LLamaWeights.CreateContext instead")]
+        public LLamaContext(IModelParams @params, string encoding = "UTF-8", ILLamaLogger? logger = null)
         {
+            Params = @params;
+
             _logger = logger;
-            this.Params = Params;
             _encoding = Encoding.GetEncoding(encoding);
-            _logger?.Log(nameof(LLamaModel), $"Initializing LLama model with params: {this.Params}", ILLamaLogger.LogLevel.Info);
-            _ctx = Utils.InitLLamaContextFromModelParams(this.Params);
-            ContextSize = NativeApi.llama_n_ctx(_ctx);
+
+            _logger?.Log(nameof(LLamaContext), $"Initializing LLama model with params: {this.Params}", ILLamaLogger.LogLevel.Info);
+            _ctx = Utils.InitLLamaContextFromModelParams(Params);
+        }
+
+        internal LLamaContext(SafeLLamaContextHandle nativeContext, IModelParams @params, Encoding encoding, ILLamaLogger? logger = null)
+        {
+            Params = @params;
+
+            _logger = logger;
+            _encoding = encoding;
+            _ctx = nativeContext;
+        }
+
+        /// <summary>
+        /// Create a new LLamaContext for the given LLamaWeights
+        /// </summary>
+        /// <param name="model"></param>
+        /// <param name="params"></param>
+        /// <param name="encoding"></param>
+        /// <param name="logger"></param>
+        /// <exception cref="ObjectDisposedException"></exception>
+        public LLamaContext(LLamaWeights model, IModelParams @params, Encoding encoding, ILLamaLogger? logger = null)
+        {
+            if (model.NativeHandle.IsClosed)
+                throw new ObjectDisposedException("Cannot create context, model weights have been disposed");
+
+            Params = @params;
+
+            _logger = logger;
+            _encoding = encoding;
+
+            using var pin = @params.ToLlamaContextParams(out var lparams);
+            _ctx = SafeLLamaContextHandle.Create(model.NativeHandle, lparams);
+        }
+
+        /// <summary>
+        /// Create a copy of the current state of this context
+        /// </summary>
+        /// <returns></returns>
+        public LLamaContext Clone()
```

Comment:
Shall we let `LLamaContext` implement `ICloneable` since the `Clone` API has already been provided?

Reply:
I'd prefer not to, the ICloneable interface isn't very well specified (e.g. is it a deep copy or a shallow copy). In fact the Microsoft documentation on `ICloneable` (https://learn.microsoft.com/en-us/dotnet/api/system.icloneable?view=net-7.0) recommends that ICloneable is _not_ implemented in public APIs:

> Because callers of [Clone()](https://learn.microsoft.com/en-us/dotnet/api/system.icloneable.clone?view=net-7.0#system-icloneable-clone) cannot depend on the method performing a predictable cloning operation, we recommend that [ICloneable](https://learn.microsoft.com/en-us/dotnet/api/system.icloneable?view=net-7.0) not be implemented in public APIs.

Reply:
Got it, you're amazing!